### ISSUE TITLE
The need to pass an explicit context? 

### ISSUE THREAD

**Issue by elizarov** [andrewoma/coroutine-jdbc#1]
The current use for transactions in the example looks like this:

```
transaction(db) { tx ->
   dao.insert(tx, kate)
   assertThat(dao.findActresses(tx).count()).isEqualTo(1)
}
```

While the variable above is named `tx`, it is actually represented by the generic type `CoroutineContext`, which is suboptimal.

Have you considered the following alternative usage that is similar to the way thread-based transactions are usually implemented?

```
transaction(db) { // context is implicit, the block has no arguments
   dao.insert(kate)
   assertThat(dao.findActresses().count()).isEqualTo(1)
}
```

The difference is that we don't explicitly pass context into `insert` and other `dao` functions. Instead, we define the following helper function:

```
import kotlin.coroutines.expertimental.intrinsics.suspendCororoutineOrReturn

suspend fun coroutineContext(): CoroutineContext = suspendCororoutineOrReturn { it.context }
```

and use this function inside DAO to retrieve the current context (with transaction if it was started) whenever we need it.

See also this issue: https://youtrack.jetbrains.com/issue/KT-17609

**Comment by andrewoma** [comment]
That's exactly what I was searching for - thanks for that!

I looked into CoroutineScope but couldn't force it into what I needed. 

I'll rework the design tomorrow (Australia time)

**Comment by andrewoma** [comment]
Fixed in c91db746501f7b739bbebfe501e22861ecc471f8
